### PR TITLE
optimizer: typeInferrer supports more expression.

### PR DESCRIPTION
**Pull request by coocood**
This is require for implementing prepared statement, because binary protocol depends more
on the result field type to decode value, we have to correctly set the result field type.

For statement like 'select ?', the type of the field is unknown until we execute the statement with argument,
If the field type of  parameter marker `?' is not set properly, client will not be able to read the value.

### CODE REVIEW

File: optimizer/typeinferer.go [hunk 1]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
+				xTp = mysql.TypeDecimal
+			}
+			x.Type = types.NewFieldType(xTp)
+		}
+	}
+}
+
+func mergeArithType(a, b byte) byte {
+	switch a {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble:
```
Comment (siddontang) [inline optimizer/typeinferer.go]: no TypeFloat?

Comment (coocood) [inline optimizer/typeinferer.go]: Oh, I missed this type.

File: util/types/field_type.go [hunk 2]
```diff
@@ -142,3 +142,47 @@ func IsDataItem(d interface{}) bool {
...
+	case mysql.Duration:
+		return NewFieldType(mysql.TypeDuration)
+	case mysql.Decimal:
+		return NewFieldType(mysql.TypeNewDecimal)
+	case mysql.Enum:
+		return NewFieldType(mysql.TypeEnum)
+	case mysql.Set:
+		return NewFieldType(mysql.TypeSet)
+	case *DataItem:
+		return value.(*DataItem).Type
```
Comment (siddontang) [inline util/types/field_type.go]: can use x.Type directly.

File: util/types/field_type.go [hunk 3]
```diff
@@ -142,3 +142,47 @@ func IsDataItem(d interface{}) bool {
...
+	}
+	return NewFieldType(mysql.TypeNull)
```
Comment (siddontang) [inline util/types/field_type.go]: why null but not error?

File: ast/expressions.go [hunk 4]
```diff
@@ -59,46 +57,13 @@ type ValueExpr struct {
...
-	switch x := value.(type) {
-	case nil:
-		ve.Type = types.NewFieldType(mysql.TypeNull)
-	case bool, int64, int:
-		ve.Type = types.NewFieldType(mysql.TypeLonglong)
-	case uint64:
-		ve.Type = types.NewFieldType(mysql.TypeLonglong)
-		ve.Type.Flag |= mysql.UnsignedFlag
-	case string, UnquoteString:
+	if _, ok := value.(UnquoteString); ok {
```
Comment (siddontang) [inline ast/expressions.go]: why using value but not ve.Date? 
why handing UnquoteString alone?

File: optimizer/typeinferer.go [hunk 5]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
-			if val.Column.ID == 0 && val.Expr.GetType() != nil {
-				val.Column.FieldType = *(val.Expr.GetType())
+		v.selectStmt(x)
+	case *ast.ParamMarkerExpr:
+		x.SetType(types.DefaultTypeForValue(x.GetValue()))
+	case *ast.BinaryOperationExpr:
+		v.binaryOperation(x)
+	case *ast.UnaryOperationExpr:
+		v.unaryOperation(x)
+	case *ast.BetweenExpr:
```
Comment (siddontang) [inline optimizer/typeinferer.go]: Putting following together may be better? like
```
case *ast.BetweenExpr, *ast.CompareSubQueryExpr, ...:
   // set TypeLongLong
```

File: optimizer/typeinferer.go [hunk 6]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.LT, opcode.LE, opcode.GE, opcode.GT, opcode.EQ, opcode.NE, opcode.NullEQ:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.RightShift, opcode.LeftShift, opcode.And, opcode.Or, opcode.Xor:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+		x.Type.Flag |= mysql.UnsignedFlag
+	case opcode.IntDiv:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.Plus, opcode.Minus, opcode.Mul, opcode.Mod:
+		if x.L.GetType() != nil && x.R.GetType() != nil {
```
Comment (siddontang) [inline optimizer/typeinferer.go]: how about Type = nil ?

File: optimizer/typeinferer.go [hunk 7]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
+		x.SetType(x.Expr.GetType())
+		// TODO: handle all expression types.
+	}
+	return in, true
+}
+
+func (v *typeInferrer) selectStmt(x *ast.SelectStmt) {
+	rf := x.GetResultFields()
+	for _, val := range rf {
+		if val.Column.ID == 0 && val.Expr.GetType() != nil {
```
Comment (qiuyesuifeng) [inline optimizer/typeinferer.go]: Can u show an example for when is val.Column.ID = 0?

File: optimizer/typeinferer.go (hunk 6 above)
Comment (qiuyesuifeng) [inline optimizer/typeinferer.go]: +1, also confused.

File: optimizer/typeinferer.go [hunk 8]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
+		x.Type.Flag |= mysql.UnsignedFlag
+	case opcode.IntDiv:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.Plus, opcode.Minus, opcode.Mul, opcode.Mod:
+		if x.L.GetType() != nil && x.R.GetType() != nil {
+			xTp := mergeArithType(x.L.GetType().Tp, x.R.GetType().Tp)
+			x.Type = types.NewFieldType(xTp)
+			leftUnsigned := x.L.GetType().Flag & mysql.UnsignedFlag
+			rightUnsigned := x.R.GetType().Flag & mysql.UnsignedFlag
+			// If both operand is unsigned, result is unsigned.
```
Comment (qiuyesuifeng) [inline optimizer/typeinferer.go]: If both operand are unsigned

File: optimizer/typeinferer.go (hunk 5 above)
Comment (coocood) [inline optimizer/typeinferer.go]: If we put multiple type in a case, we can't use the real typed `x`, we can only use the `in`,
`in` is `ast.Node`, not `ast.ExprNode`, it doesn't have `SetType` method.

File: optimizer/typeinferer.go (hunk 7 above)
Comment (coocood) [inline optimizer/typeinferer.go]: `Conlumn.ID = 0` means the column is not a real column from table, but a generated column for result field.

File: optimizer/typeinferer.go (hunk 6 above)
Comment (coocood) [inline optimizer/typeinferer.go]: `Type == nil` means the expression type is not handled for now, we can't infer the expression type.

File: ast/expressions.go (hunk 4 above)
Comment (coocood) [inline ast/expressions.go]: `UnquoteString` is only used for wildcard symbol `*` will not be wrapped in a `DataItem` so it is ok to use value direclty.
And `UnquoteString` is a type in `ast`, package `types` can not handle this type.

File: util/types/field_type.go (hunk 3 above)
Comment (coocood) [inline util/types/field_type.go]: I'll change it to log an error and return nil.

**Comment by coocood** [comment]
PTAL @siddontang @qiuyesuifeng

File: optimizer/typeinferer.go (hunk 6 above)
Comment (siddontang) [inline optimizer/typeinferer.go]: then how do we infer the type later?

File: optimizer/typeinferer.go (hunk 7 above)
Comment (shenli) [inline optimizer/typeinferer.go]: So we need a comment here or replace '0' with a const.

File: optimizer/typeinferer.go [hunk 9]
```diff
@@ -48,14 +51,111 @@ func (v *typeInferrer) Leave(in ast.Node) (out ast.Node, ok bool) {
...
+
+func (v *typeInferrer) unaryOperation(x *ast.UnaryOperationExpr) {
+	switch x.Op {
+	case opcode.Not:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+	case opcode.BitNeg:
+		x.Type = types.NewFieldType(mysql.TypeLonglong)
+		x.Type.Flag |= mysql.UnsignedFlag
+	case opcode.Plus:
+		x.Type = x.V.GetType()
```
Comment (shenli) [inline optimizer/typeinferer.go]: Should consider the same things as opcode.Minus?

Comment (coocood) [inline optimizer/typeinferer.go]: Plus works differently than Minus, `select +'abc'` returns the original varchar type.

File: optimizer/typeinferer.go (hunk 6 above)
Comment (coocood) [inline optimizer/typeinferer.go]: This implementation is not complete, when we handle every expression, then we can be sure every expression's type is not nil.

**Comment by shenli** [comment]
LGTM

File: optimizer/typeinferer.go (hunk 6 above)
Comment (ngaut) [inline optimizer/typeinferer.go]: So we need to add more todos ?

Comment (coocood) [inline optimizer/typeinferer.go]: The TODO in `typeinferer.go:79` can cover this.

File: util/types/field_type.go [hunk 10]
```diff
@@ -142,3 +143,48 @@ func IsDataItem(d interface{}) bool {
...
+	case mysql.Decimal:
+		return NewFieldType(mysql.TypeNewDecimal)
+	case mysql.Enum:
+		return NewFieldType(mysql.TypeEnum)
+	case mysql.Set:
+		return NewFieldType(mysql.TypeSet)
+	case *DataItem:
+		return x.Type
+	}
+	log.Errorf("Unknown value type %T for default field type.", value)
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: If an unknown type found, return nil is safe?

File: optimizer/typeinferer_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		{"1 in (2, 3)", mysql.TypeLonglong},
+		{"'abc' like 'abc'", mysql.TypeLonglong},
+		{"'abc' rlike 'abc'", mysql.TypeLonglong},
+		{"(1+1)", mysql.TypeLonglong},
+	}
+	for _, ca := range cases {
+		ctx := testKit.Se.(context.Context)
+		stmts, err := tidb.Parse(ctx, "select "+ca.expr+" from t")
+		c.Assert(err, IsNil)
+		c.Assert(len(stmts), Equals, 1)
```
Comment (qiuyesuifeng) [inline optimizer/typeinferer_test.go]: Please using `Haslen`.

File: util/types/field_type.go (hunk 10 above)
Comment (coocood) [inline util/types/field_type.go]: Type nil is checked before use. It will not cause panic.

**Pull request merged**